core/src/main/scala/org/apache/spark/rpc/netty/Inbox.scala

```diff
@@ -0,0 +1,229 @@
+    val endpoint: RpcEndpoint) extends Logging {
+
+  private val supportConcurrent = !endpoint.isInstanceOf[ThreadSafeRpcEndpoint]
+
+  @GuardedBy("this")
+  protected val messages = new LinkedList[InboxMessage]()
+
+  @GuardedBy("this")
+  private var stopped = false
+
+  @GuardedBy("this")
+  private var enableConcurrent = false
+
+  @GuardedBy("this")
+  private var workerCount = 0
+
+  // OnStart should be the first message to process
+  synchronized {
+    messages.add(OnStart)
+  }
+
+  def process(dispatcher: Dispatcher): Boolean = {
+    var message: InboxMessage = null
+    synchronized {
+      if (!enableConcurrent && workerCount != 0) {
+        return false
+      }
+      message = messages.poll()
+      if (message != null) {
+        workerCount += 1
+      } else {
+        return false
+      }
+    }
+    var skipFinally = false
+    try {
+      while (true) {
+        safelyCall(endpoint) {
+          message match {
+            case ContentMessage(_sender, content, needReply, context) =>
+              val pf: PartialFunction[Any, Unit] =
+                if (needReply) {
+                  endpoint.receiveAndReply(context)
+                } else {
+                  endpoint.receive
+                }
+              try {
+                pf.applyOrElse[Any, Unit](content, { msg =>
+                  throw new SparkException(s"Unmatched message $message from ${_sender}")
+                })
+                if (!needReply) {
+                  context.finish()
+                }
+              } catch {
+                case NonFatal(e) =>
+                  if (needReply) {
+                    // If the sender asks a reply, we should send the error back to the sender
+                    context.sendFailure(e)
+                  } else {
+                    context.finish()
+                    throw e
+                  }
+              }
+
+            case OnStart => {
+              endpoint.onStart()
+              if (supportConcurrent) {
+                synchronized {
+                  enableConcurrent = true
+                }
+              }
+            }
+            case OnStop =>
+              dispatcher.removeRpcEndpointRef(endpoint)
+              endpoint.onStop()
+              assert(isEmpty, "OnStop should be the last message")
+              return true
+            case Associated(remoteAddress) =>
+              endpoint.onConnected(remoteAddress)
+            case Disassociated(remoteAddress) =>
+              endpoint.onDisconnected(remoteAddress)
+            case AssociationError(cause, remoteAddress) =>
+              endpoint.onNetworkError(cause, remoteAddress)
+          }
+        }
+
+        synchronized {
+          // "enableConcurrent" will be set to false after `onStop` is called, so we should check it
+          // every time.
+          if (!enableConcurrent && workerCount != 1) {
+            // If we are not the only one worker, exit
+            skipFinally = true
```

Comment:
Removed `skipFinally`. I was worried about the exception. But since `safetyCall` has handled it, `skipFinally` is not necessary.